x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java

```diff
@@ -62,63 +85,159 @@ public interface SecurityMigration {
     }
 
     public static final Integer ROLE_METADATA_FLATTENED_MIGRATION_VERSION = 1;
+    public static final Integer CLEANUP_ROLE_MAPPING_DUPLICATES_MIGRATION_VERSION = 2;
+    private static final Logger logger = LogManager.getLogger(SecurityMigration.class);
 
     public static final TreeMap<Integer, SecurityMigration> MIGRATIONS_BY_VERSION = new TreeMap<>(
-        Map.of(ROLE_METADATA_FLATTENED_MIGRATION_VERSION, new SecurityMigration() {
-            private static final Logger logger = LogManager.getLogger(SecurityMigration.class);
-
-            @Override
-            public void migrate(SecurityIndexManager indexManager, Client client, ActionListener<Void> listener) {
-                BoolQueryBuilder filterQuery = new BoolQueryBuilder().filter(QueryBuilders.termQuery("type", "role"))
-                    .mustNot(QueryBuilders.existsQuery("metadata_flattened"));
-                SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder().query(filterQuery).size(0).trackTotalHits(true);
-                SearchRequest countRequest = new SearchRequest(indexManager.getConcreteIndexName());
-                countRequest.source(searchSourceBuilder);
-
-                client.search(countRequest, ActionListener.wrap(response -> {
-                    // If there are no roles, skip migration
-                    if (response.getHits().getTotalHits().value > 0) {
-                        logger.info("Preparing to migrate [" + response.getHits().getTotalHits().value + "] roles");
-                        updateRolesByQuery(indexManager, client, filterQuery, listener);
+        Map.of(
+            ROLE_METADATA_FLATTENED_MIGRATION_VERSION,
+            new RoleMetadataFlattenedMigration(),
+            CLEANUP_ROLE_MAPPING_DUPLICATES_MIGRATION_VERSION,
+            new CleanupRoleMappingDuplicatesMigration()
+        )
+    );
+
+    public static class RoleMetadataFlattenedMigration implements SecurityMigration {
```

Comment:
No changes here, just moved this to a named class since I wanted `CleanupRoleMappingDuplicatesMigration` to be a named class for testing purposes and it felt weird to have one anonymous migration and one that is not.